qualtran/_infra/data_types.py

```diff
@@ -810,6 +810,18 @@ def assert_valid_classical_val_array(
         if np.any(val_array >= 2**self.bitsize):
             raise ValueError(f"Too-large classical values encountered in {debug_str}")
 
+    def montgomery_inverse(self, xm, n, p):
+        return ((pow(xm, p - 2, p)) * pow(2, 2 * n, p)) % p
+
+    def montgomery_product(self, xm, ym, n, p):
+        return (xm * ym * pow(2, n * (p - 2), p)) % p
```

Comment:
can you write some unit tests for these operations? maybe check that prod(x, inv(x)) = 1